oci/cache/cache.go

```diff
@@ -0,0 +1,159 @@
+// Ideally we can merge this behavior into the oras-go library - https://github.com/oras-project/oras-go/issues/881
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package cache is used to create a ReadOnlyTarget from an existing target and an oci.Store
+package cache
+
+import (
+	"context"
+	"io"
+	"sync"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/registry"
+)
+
+type closer func() error
+
+func (fn closer) Close() error {
+	return fn()
+}
+
+// Cache target struct.
+type target struct {
+	oras.ReadOnlyTarget
+	cache content.Storage
+}
+
+// New generates a new target storage with caching.
+func New(source oras.ReadOnlyTarget, cache content.Storage) oras.ReadOnlyTarget {
+	t := &target{
+		ReadOnlyTarget: source,
+		cache:          cache,
+	}
+	if refFetcher, ok := source.(registry.ReferenceFetcher); ok {
+		return &referenceTarget{
+			target:           t,
+			ReferenceFetcher: refFetcher,
+		}
+	}
+	return t
+}
+
+// Fetch fetches the content identified by the descriptor.
+func (t *target) Fetch(ctx context.Context, target ocispec.Descriptor) (io.ReadCloser, error) {
+	rc, err := t.cache.Fetch(ctx, target)
+	if err == nil {
+		// Fetch from cache
+		return rc, nil
+	}
+
+	if rc, err = t.ReadOnlyTarget.Fetch(ctx, target); err != nil {
+		return nil, err
+	}
+
+	// Fetch from origin with caching
+	return t.cacheReadCloser(ctx, rc, target), nil
+}
+
+func (t *target) cacheReadCloser(ctx context.Context, rc io.ReadCloser, target ocispec.Descriptor) io.ReadCloser {
+	pr, pw := io.Pipe()
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	var pushErr error
+	go func() {
+		defer wg.Done()
+		pushErr = t.cache.Push(ctx, target, pr)
+		if pushErr != nil {
+			pr.CloseWithError(pushErr)
+		}
+	}()
+
+	return struct {
+		io.Reader
+		io.Closer
+	}{
+		Reader: io.TeeReader(rc, pw),
+		Closer: closer(func() error {
+			rcErr := rc.Close()
+			if err := pw.Close(); err != nil {
+				return err
+			}
+			wg.Wait()
+			if pushErr != nil {
+				return pushErr
+			}
+			return rcErr
+		}),
+	}
+}
+
+// Exists returns true if the described content exists.
+func (t *target) Exists(ctx context.Context, desc ocispec.Descriptor) (bool, error) {
+	exists, err := t.cache.Exists(ctx, desc)
+	if err == nil && exists {
+		return true, nil
+	}
+	return t.ReadOnlyTarget.Exists(ctx, desc)
+}
```

Comment:
some cache errors will not be reported, this is intended behavior but is worth noting